typify-impl/src/merge.rs

```diff
@@ -138,15 +160,67 @@ fn try_merge_schema(a: &Schema, b: &Schema, defs: &BTreeMap<RefKey, Schema>) ->
             }
         }
 
-        (Schema::Object(aa), Schema::Object(bb)) => Ok(merge_schema_object(aa, bb, defs)?.into()),
+        (Schema::Object(aa), Schema::Object(bb)) => Ok(merge_schema_object(aa, bb, defs)
+            .map_err(|source| SchemaMergeError::ObjectSchemaMerge {
+                source: Box::new(source),
+                a: aa.clone(),
+                b: bb.clone(),
+            })?
+            .into()),
     }
 }
 
+#[derive(Error, Debug)]
+pub(crate) enum ObjectSchemaMergeError {
+    #[error("Error when trying to merge the two instance types {a:?} and {b:?}")]
```

Comment:
How about we use the term "incompatible" here and throughout?

```suggestion
    #[error("Incompatible instance types {a:?} and {b:?}")]
```

Reply:
Done